File: packages/companion/swift/Sources/BridgethingCompanion/VoiceController.swift
import BridgethingSchema
import Foundation

public actor VoiceController {
    public struct Config: Sendable {
        public let useFastPath: Bool
        public let rejection: NluRejectionPolicy

        public init(useFastPath: Bool = true, rejection: NluRejectionPolicy = .init()) {
            self.useFastPath = useFastPath
            self.rejection = rejection
        }
    }

    public enum Stage: String, Sendable {
        case fastPath
        case model
        case rejectedNoIntent
        case rejectedClarify
        case noModel

        var wire: NluStage {
            switch self {
            case .fastPath: return .fastPath
            case .model: return .model
            case .rejectedNoIntent: return .rejectedNoIntent
            case .rejectedClarify: return .rejectedClarify
            case .noModel: return .noModel
            }
        }
    }

    public struct Resolution: Sendable {
        public let resolved: NluResolvedIntent
        public let stage: Stage

        public init(resolved: NluResolvedIntent, stage: Stage) {
            self.resolved = resolved
            self.stage = stage
        }
    }

    public enum ControllerError: Error, CustomStringConvertible {
        case inferenceFailed(Error)

        public var description: String {
            switch self {
            case let .inferenceFailed(err): return "nlu inference failed: \(err)"
            }
        }
    }

    private let client: (any NluInferring)?
    private let config: Config

    public init(client: (any NluInferring)? = nil, config: Config = Config()) {
        self.client = client
        self.config = config
    }

    public func resolve(transcript: String) async throws -> Resolution {
        let trimmed = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return Resolution(
                resolved: NluPrediction(intent: NluIntentCatalog.noIntent, transcript: transcript).toWire(),
                stage: .rejectedNoIntent
            )
        }

        if config.useFastPath, let hit = NluFastPath.match(trimmed) {
            let pred = NluPrediction(intent: hit.intent, slots: hit.slots, transcript: transcript)
            return Resolution(resolved: pred.toWire(), stage: .fastPath)
        }

        guard let client else {
            return Resolution(
                resolved: NluPrediction(intent: NluIntentCatalog.noIntent, transcript: transcript).toWire(),
                stage: .noModel
            )
        }

        let output: NluInferenceOutput
        do {
            output = try await client.infer(transcript: trimmed)
        } catch {
            throw ControllerError.inferenceFailed(error)
        }

        switch try NluRejection.evaluate(output, policy: config.rejection) {
        case .noIntent:
            return Resolution(
                resolved: NluPrediction(intent: NluIntentCatalog.noIntent, transcript: transcript).toWire(),
                stage: .rejectedNoIntent
            )

        case let .clarify(alternates):
            let pred = NluPrediction(
                intent: NluIntentCatalog.clarify,
                transcript: transcript,
                alternates: alternates.map { NluAlternate(intent: $0, slots: nil) }
            )
            return Resolution(resolved: pred.toWire(), stage: .rejectedClarify)

        case let .accept(intent):
            let pred = NluPrediction(intent: intent, slots: output.slots, transcript: transcript)
            return Resolution(resolved: pred.toWire(), stage: .model)
        }
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/VoiceController.kt
package com.bridgething.companion

import com.bridgething.schema.NluAlternate
import com.bridgething.schema.NluResolvedIntent
import com.bridgething.schema.NluStage
import kotlinx.coroutines.CancellationException

class VoiceController(
    private val client: NluInferring? = null,
    private val config: Config = Config(),
) {
    data class Config(
        val useFastPath: Boolean = true,
        val rejection: NluRejectionPolicy = NluRejectionPolicy(),
    )

    data class Resolution(val resolved: NluResolvedIntent, val stage: NluStage)

    class InferenceFailed(cause: Throwable) : Exception("nlu inference failed: $cause", cause)

    suspend fun prewarm() {
        (client as? NluPrewarmable)?.prewarm()
    }

    suspend fun resolve(transcript: String): Resolution {
        val trimmed = transcript.trim()
        if (trimmed.isEmpty()) return noIntent(transcript, NluStage.RejectedNoIntent)

        if (config.useFastPath) {
            NluFastPath.match(trimmed)?.let { hit ->
                val pred = NluPrediction(intent = hit.intent, transcript = transcript, slots = hit.slots)
                return Resolution(pred.toWire(), NluStage.FastPath)
            }
        }

        val client = client ?: return noIntent(transcript, NluStage.NoModel)

        val output = try {
            client.infer(trimmed)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            throw InferenceFailed(e)
        }

        return when (val outcome = NluRejection.evaluate(output, config.rejection)) {
            is NluRejectionOutcome.NoIntent -> noIntent(transcript, NluStage.RejectedNoIntent)

            is NluRejectionOutcome.Clarify -> {
                val pred = NluPrediction(
                    intent = NluIntentCatalog.CLARIFY,
                    transcript = transcript,
                    alternates = outcome.alternates.map { NluAlternate(intent = it, slots = null) },
                )
                Resolution(pred.toWire(), NluStage.RejectedClarify)
            }

            is NluRejectionOutcome.Accept -> {
                val pred = NluPrediction(
                    intent = outcome.intent,
                    transcript = transcript,
                    slots = NluMutableSlots.fromWire(output.slots),
                )
                Resolution(pred.toWire(), NluStage.Model)
            }
        }
    }

    private fun noIntent(transcript: String, stage: NluStage): Resolution =
        Resolution(
            NluPrediction(intent = NluIntentCatalog.NO_INTENT, transcript = transcript).toWire(),
            stage,
        )
}
